src/thumbnailer_near_lossless.cc

```diff
@@ -0,0 +1,185 @@
+      // Binary search for near-lossless's pre-processing value.
+      while (min_near_ll <= max_near_ll) {
+        int mid_near_ll = (min_near_ll + max_near_ll) / 2;
+        frame.config.near_lossless = mid_near_ll;
+        CHECK_THUMBNAILER_STATUS(
+            GetPictureStats(curr_ind, &new_size, &new_psnr));
+        if (anim_size - curr_size + new_size <= byte_budget_) {
+          if (new_psnr > curr_psnr) {
+            final_near_ll = mid_near_ll;
+            frame.encoded_size = new_size;
+            frame.final_psnr = new_psnr;
+            frame.final_quality = frame.config.quality;
+            frame.near_lossless = true;
+            anim_size = anim_size - curr_size + new_size;
+            curr_size = new_size;
+            curr_psnr = new_psnr;
+          }
+          min_near_ll = mid_near_ll + 1;
+        } else {
+          max_near_ll = mid_near_ll - 1;
+        }
+      }
+    }
+
+    std::cerr << final_near_ll << " ";
+
+    if (final_near_ll == -1) {
+      frame.config.lossless = 0;
+      frame.config.quality = frame.final_quality;
+    } else {
+      frame.config.near_lossless = final_near_ll;
+    }
+
+    ++curr_ind;
+  }
+
+  std::cerr << std::endl;
+
+  WebPDataClear(webp_data);
+  CHECK_THUMBNAILER_STATUS(GenerateAnimationNoBudget(webp_data));
+
+  return (webp_data->size > 0) ? kOk : kByteBudgetError;
+}
+
+Thumbnailer::Status Thumbnailer::NearLosslessEqualPreProcessing(
+    WebPData* const webp_data) {
+  const int num_frames = frames_.size();
+
+  // Encode frames following the ascending order of frame sizes.
+  std::pair<int, int> encoding_order[num_frames];
+  for (int i = 0; i < num_frames; ++i) {
+    encoding_order[i] = std::make_pair(frames_[i].encoded_size, i);
+  }
+  std::sort(encoding_order, encoding_order + num_frames,
+            [](const std::pair<int, int>& x, const std::pair<int, int>& y) {
+              return x.first < y.first;
+            });
+
+  // Vector of frames encoded with near-lossless preprocessing 0.
+  std::vector<int> near_ll_frames;
+  int anim_size = GetAnimationSize(webp_data);
+  // Find the maximum number of frames that can be encoded with near-lossless
+  // preprocessing 0.
+  for (int i = 0; i < num_frames; ++i) {
+    const int curr_ind = encoding_order[i].second;
+    frames_[curr_ind].config.lossless = 1;
+    frames_[curr_ind].config.quality = 90;
+    frames_[curr_ind].config.near_lossless = 0;
+    int new_size;
+    float new_psnr;
+    CHECK_THUMBNAILER_STATUS(GetPictureStats(curr_ind, &new_size, &new_psnr));
+    const int new_anim_size =
+        anim_size - frames_[curr_ind].encoded_size + new_size;
+    if (new_psnr >= frames_[curr_ind].final_psnr &&
+        new_anim_size <= byte_budget_) {
+      anim_size = new_anim_size;
+      near_ll_frames.push_back(curr_ind);
+      frames_[curr_ind].encoded_size = new_size;
+      frames_[curr_ind].final_psnr = new_psnr;
+      frames_[curr_ind].final_quality = 90;
```

Comment:
`frames_[curr_ind].final_quality = frames_[curr_ind].config.quality;` for safety.
Or should it be `frames_[curr_ind].config.near_lossless` actually, otherwise the final `near_lossless` value is not remembered right?

Reply:
That's right, the final `frames_[curr_ind].config.near_lossless` is not remembered, it's already stored in the `final_near_ll` variable.

Reply:
Then maybe `final_near_ll` should outlive `NearLosslessEqualPreProcessing()`? Setting all `frames_[...].config.near_lossless = final_near_ll;` around line 180 would work I guess, similar to what is done at line 75

Reply:
Thanks a lot, you've helped me find a bug!